prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,79 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
 		wc->byte_len : 0;
 }
 
-static int fi_ibv_match_ep_id(struct slist_entry *entry, const void *ep_id)
+static inline int fi_ibv_match_ep_id(struct slist_entry *entry,
+				     const void *ep_id)
 {
-	struct fi_ibv_msg_epe *epe = container_of(entry, struct fi_ibv_msg_epe, entry);
+	struct fi_ibv_msg_epe *epe =
+		container_of(entry, struct fi_ibv_msg_epe, entry);
 
-	if (epe->ep->ep_id == (uint64_t) ep_id)
-		return 1;
+	return (epe->ep->ep_id == (uint64_t)ep_id);
+}
 
-	return 0;
+static inline int fi_ibv_match_ep(struct slist_entry *entry,
+				  const void *cur_ep)
+{
+	struct fi_ibv_wce *wce =
+		container_of(entry, struct fi_ibv_wce, entry);
+	struct fi_ibv_wre *wre =
+		(struct fi_ibv_wre *)
+			((struct fi_context *)wce->wc.wr_id)->internal[0];
+
+	return (wre->ep == cur_ep);
 }
 
 /* Must call with cq->lock held */
 ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
 {
 	struct fi_ibv_msg_epe *epe;
+	struct fi_ibv_wre *wre;
 	struct slist_entry *entry;
 	ssize_t ret;
 
 	ret = ibv_poll_cq(cq->cq, 1, wc);
 	if (ret <= 0)
 		return ret;
 
-	if (wc->opcode == IBV_WC_RECV || wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM)
-		return ret;
+	if ((wc->opcode == IBV_WC_RECV) ||
+	    (wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM))
+		goto done;
 
 	/* TODO Handle the case when app posts a send with same wr_id */
 	if ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)
-		return ret;
+		goto done;
 
-	entry = slist_remove_first_match(&cq->ep_list, fi_ibv_match_ep_id, (void *)wc->wr_id);
+	entry = slist_remove_first_match(&cq->ep_list,
+					 fi_ibv_match_ep_id,
+					 (void *)wc->wr_id);
 	if (!entry) {
 		VERBS_WARN(FI_LOG_CQ, "No matching EP for :"
 			   "given signaled send completion\n");
 		return -FI_EOTHER;
 	}
 	epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 	ofi_atomic_sub32(&epe->ep->unsignaled_send_cnt,
-			VERBS_SEND_SIGNAL_THRESH(epe->ep));
+			 VERBS_SEND_SIGNAL_THRESH(epe->ep));
 	ofi_atomic_dec32(&epe->ep->comp_pending);
 	util_buf_release(cq->epe_pool, epe);
 
 	return 0;
+
+done:
+	wre = (struct fi_ibv_wre *)
+		((struct fi_context *)wc->wr_id)->internal[0];
+	if (wre->ep && (wc->status == IBV_WC_WR_FLUSH_ERR)) {
```

Comment:
if we get `IBV_WC_WR_FLUS_ER` it assumes that no `RDMA_CM_EVENT_DISCONNECTED` is received and we initiate disconnect to remote side

Reply:
>if we get IBV_WC_WR_FLUS_ER it assumes that no RDMA_CM_EVENT_DISCONNECTED is received and we initiate disconnect to remote side

Does this happen when the remote QP is closed? I was of the idea that we should get a 'disconnect' event when remote QP is closed.

> we should implement the releasing of fi_ibv_wre buffers and removing them from the wre_list for the current EP when RDMA_CM_EVENT_DISCONNECTED is received or EP is closing

👍  The app should close the EP on receiving a disconnect / shutdown event. So releasing the buffers on EP close should be sufficient.

Reply:
> Does this happen when the remote QP is closed? I was of the idea that we should get a 'disconnect' event when remote QP is closed.

Yes, I was of the the same idea, but it's not so. We should not rely on the underlying driver (maybe it's just a behavior of the libnes), the verbs provider will manage it in itself